sdk/webpubsub/azure-messaging-webpubsub-client/src/main/java/com/azure/messaging/webpubsub/client/implementation/websocket/WebSocketClientHandler.java

```diff
@@ -46,80 +52,136 @@ ChannelFuture handshakeFuture() {
     }
 
     @Override
-    public void handlerAdded(ChannelHandlerContext ctx) {
-        handshakeFuture = ctx.newPromise();
+    public void handlerAdded(ChannelHandlerContext context) {
+        handshakeFuture = context.newPromise();
+        compositeByteBuf = context.alloc().compositeBuffer();
     }
 
     @Override
-    public void channelActive(ChannelHandlerContext ctx) {
-        handshaker.handshake(ctx.channel());
+    public void handlerRemoved(ChannelHandlerContext ctx) {
+        try {
+            final BinaryData data = BinaryData.fromListByteBuffer(Arrays.asList(compositeByteBuf.nioBuffers()));
+            final String collected = data.toString();
+            final WebPubSubMessage deserialized = messageDecoder.decode(collected);
+
+            messageHandler.accept(deserialized);
+        } finally {
+            compositeByteBuf.release();
+            compositeByteBuf.clear();
+        }
     }
 
     @Override
-    protected void channelRead0(ChannelHandlerContext ctx, Object msg) {
-        Channel ch = ctx.channel();
+    public void channelActive(ChannelHandlerContext context) {
+        handshaker.handshake(context.channel());
+    }
+
+    @Override
+    protected void channelRead0(ChannelHandlerContext context, Object message) {
         if (handshakeFuture != null && !handshaker.isHandshakeComplete()) {
+            Channel channel = context.channel();
+
             try {
-                handshaker.finishHandshake(ch, (FullHttpResponse) msg);
+                handshaker.finishHandshake(channel, (FullHttpResponse) message);
                 handshakeFuture.setSuccess();
             } catch (WebSocketHandshakeException e) {
                 handshakeFuture.setFailure(e);
             }
             return;
         }
 
-        if (msg instanceof FullHttpResponse) {
-            FullHttpResponse response = (FullHttpResponse) msg;
-            throw loggerReference.get()
-                .logExceptionAsError(new IllegalStateException("Unexpected FullHttpResponse (getStatus="
-                    + response.status() + ", content=" + response.content().toString(CharsetUtil.UTF_8) + ')'));
+        if (!(message instanceof WebSocketFrame) || !processMessage(context, (WebSocketFrame) message)) {
+            loggerReference.get()
+                .atWarning()
+                .addKeyValue("messageType", message.getClass())
+                .log("Unknown message type. Skipping.");
+
+            context.fireChannelRead(message);
+        }
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        if (handshakeFuture != null && !handshakeFuture.isDone()) {
+            handshakeFuture.setFailure(cause);
         }
+        ctx.close();
+        compositeByteBuf.release();
+        compositeByteBuf.clear();
```

Comment:
If we attempt to release a `ByteBuf` (`compositeByteBuf`) instance when its `refCount` is 0, it will cause Netty to throw `IllegalReferenceCountException`. In our case, such unintentional releases may occur, e.g., if an exception is notified after the final-fragment handling in `processMessage`. 

We could move buffer release with guard to a function and reuse it in all places

```java
private static void release(CompositeByteBuf buffer) {
   if (buffer.refCount() > 0) {
       buffer.release();
       buffer.clear();
   }
}
```